keps/sig-api-machinery/3037-client-go-alternative-services/README.md

```diff
@@ -0,0 +1,590 @@
+
+
+```sh
+1103 12:30:59.066469 1558329 round_trippers.go:454] GET https://127.0.0.1:44267/api/v1/namespaces/default/pods?limit=500 200 OK in 1 milliseconds
+I1103 12:30:59.066484 1558329 round_trippers.go:460] Response Headers:
+I1103 12:30:59.066491 1558329 round_trippers.go:463]     Cache-Control: no-cache, private
+I1103 12:30:59.066502 1558329 round_trippers.go:463]     Alt-Svc: h2="10.0.0.2:6443", h2="10.0.0.3:6443", h2="10.0.0.4:6443
+```
+
+The Alternative Services RFC defines options to [implement caching](https://datatracker.ietf.org/doc/html/rfc7838#section-3.1),
+however, it doesn't seem necessary because:
+
+- kubernetes control-plane is not a high dynamic environment.
+- apiserver already has another mechanisms to warn clients using http codes.
+- if the alt-service has gone you will have a connection error, the client will expire that entry from the cache and will try another alt-svc.
+
+### Client side HA
+
+Client-go creates a layer on top of the golang net/http library abstracting the communications against the apiservers using [Requests](https://github.com/kubernetes/client-go/blob/master/rest/request.go).
+
+The golang net/http library allows to modify a request in different points:
+- RoundTripper/Transport
+- Dialer
+- DNS dialer
+
+The client-go base-URL is immutable after creation, that guarantees that Requests will not be able to modify it externally.
+
+Client-go already implements some [custom RoundTrippers](https://github.com/kubernetes/client-go/blob/master/transport/round_trippers.go) for some functionalities, like debugging, authentication, impersonation, ...
+
+The proposal is to implement a new round tripper that has a local cache with the list of available apiservers.
+This list can be created manually via configuration (comma-separated list on the Config object) or automatically via the Alt-Svc headers published.
+
+The round tripper behavior, if there are alternative services, will be:
+
+1. Prefer local (same system as the client) alternative services.
+2. Prefer the one in the original request if is listed as an alternative service.
+3. If one alternative service is chosen and succeed, stick to it, only change on network failures.
+4. If the connection against an alternative service fails, block that alternative service for a specified timeout so it will not be retried, retry against the other available alternative services, if everything fails fall back to the original host.
+
+It retries only in case of network failures, it operates only at the network L2/L3/L4 level, that guarantees the idempotency of the request and any possible conflict with others roundtrippers and the client-go retry logic.
```

Comment:
> 3 and 4 require polling the apiserver

Wont dialer be called again when app requests fail (or ping fail) and it reconnects?

Reply:
yeah, but the dialer has no access to the "alt-svc" headers, so it can not know what apiservers are healthy.
A dialer needs a loop that polls the apiservers /healthz endpoints to build the local cache or to sneak into the transport, both are bad.

Reply:
I think you don't need to check healthz. The existing logic for closing out borked HTTP/2 connections should just work.

Reply:
We have this problem in openshift very often, apiserver receiving requests when are shutting down or starting, we want to guarantee that you reach an apiserver that will be ready to server